File: xweb/static/validators.js
/**
 * XwebValidate — port client (JS) du moteur de validation déclarative
 * xdsl/validators.py, pour du feedback instantané côté navigateur (avant
 * même d'envoyer une requête) sur des données reçues en live ou la RÉPONSE
 * JSON d'un `endpoint { receive { schema: ... } }` (voir
 * xdsl/compiler.py::_write_endpoint_scaffolding et ::_write_channel_bootstrap).
 *
 * Écrit à la main, comme le reste du JS non-vendorisé de ce dépôt
 * (storage.js, live_channel.js, script anti-flash de thème) : même
 * ensemble FERMÉ de règles des deux côtés, pas une lib de validation
 * générique — voir BUILTIN_VALIDATORS dans xdsl/validators.py, la seule
 * source de vérité sur les règles qui existent.
 *
 * NE REMPLACE JAMAIS la validation SERVEUR (xdsl/validators.py::validate_dict,
 * appelée par le compilateur DSL sur tout `endpoint`/`receive`) — c'est la
 * seule qui compte pour la sécurité (docs/dsl-design.md#requêtes-réseau,
 * CLAUDE.md "server est la seule source de vérité pour la sécurité"). Ce
 * fichier n'existe QUE pour de l'UX (surligner un champ, bloquer un
 * rafraîchissement DOM sur une donnée live malformée, bloquer onsuccess/
 * l'event de refresh d'un endpoint sur une réponse JSON invalide) — un
 * client hostile peut trivialement l'ignorer ou le contourner.
 *
 * Contrat :
 *   window.XWEB_SCHEMAS[name] = {
 *     champ: [[règle, [args...]], ...]   // ex: [["required", []], ["min_length", [3]]]
 *   }
 *   — écrit par le bootstrap généré (`channel { validate: "nom" }` ou
 *   `endpoint { receive { schema: "nom" } }`), jamais à la main.
 *   window.XwebValidate(name, data, opts) lit ce registre.
 *
 * COMPOSITION — un champ dont le type n'est pas un primitif (split par
 * xdsl/parser.py::split_value_type) porte une pseudo-règle qui n'est PAS
 * dans RULES, interprétée ici pour DESCENDRE :
 *   ["ref", ["address"]]   — value doit être un objet, validé contre le
 *                            schéma "address" (erreurs préfixées `address.street`).
 *   ["list", ["phone"]]    — value doit être un tableau, chaque item validé
 *                            contre le schéma "phone" (erreurs préfixées
 *                            `phones[0].number`), OU contre un PRIMITIF
 *                            (`["list", ["string"]]`, contrôle de type par item).
 *
 * window.XwebValidate(schemaName, data, opts) -> { valid: bool, errors: {champ: [msg...]} }
 *   - opts.list === true : data est un TABLEAU, chaque item validé contre le
 *     schéma, erreurs préfixées `0.name`, ... (`receive { list: true }`).
 *   - schéma introuvable dans window.XWEB_SCHEMAS -> échoue de façon
 *     VISIBLE (console.error + valid:false), jamais un silencieux
 *     valid:true qui masquerait un nom de schéma mal tapé.
 *   - règle inconnue dans le descripteur -> pareil, console.error +
 *     valid:false (ne devrait jamais arriver : le compilateur xdsl ne
 *     génère que des règles qu'il connaît lui-même).
 */
(function () {
  "use strict";

  // Miroir exact de xdsl/validators.py::BUILTIN_VALIDATORS — même noms,
  // mêmes messages (français), mêmes règles de passthrough (une règle
  // typée string/nombre qui reçoit une valeur du mauvais type LAISSE
  // PASSER, elle ne s'applique juste pas — voir MinLength.check/MinValue.check
  // côté Python, ce n'est pas un oubli ici).
  const RULES = {
    required(value) {
      if (value === null || value === undefined) return false;
      if (typeof value === "string" && value.trim() === "") return false;
      return true;
    },
    min_length(value, [minLen]) {
      if (typeof value !== "string") return true;
      return value.length >= minLen;
    },
    max_length(value, [maxLen]) {
      if (typeof value !== "string") return true;
      return value.length <= maxLen;
    },
    min(value, [minVal]) {
      const n = Number(value);
      if (value === null || value === undefined || value === "" || Number.isNaN(n)) return true;
      return n >= minVal;
    },
    max(value, [maxVal]) {
      const n = Number(value);
      if (value === null || value === undefined || value === "" || Number.isNaN(n)) return true;
      return n <= maxVal;
    },
    email(value) {
      if (typeof value !== "string") return true;
      return /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/.test(value);
    },
    pattern(value, [regex]) {
      if (typeof value !== "string") return true;
      return new RegExp("^(?:" + regex + ")").test(value);
    },
    in(value, [allowed]) {
      return allowed.map(String).includes(String(value));
    },
  };

  const MESSAGES = {
    required: () => "Ce champ est obligatoire",
    min_length: ([n]) => `Minimum ${n} caractère${n > 1 ? "s" : ""}`,
    max_length: ([n]) => `Maximum ${n} caractère${n > 1 ? "s" : ""}`,
    min: ([n]) => `La valeur minimale est ${n}`,
    max: ([n]) => `La valeur maximale est ${n}`,
    email: () => "Adresse email invalide",
    pattern: () => "La valeur ne correspond pas au format attendu",
    in: ([allowed]) => `Valeur invalide — autorisées : ${allowed.join(", ")}`,
  };

  // Miroir du wording de type de pydantic_bridge._TYPE_ERRORS (le serveur,
  // source de vérité — une liste reçoit `["list", [item]]` via la pseudo-
  // règle de _schema_to_json) — erreurs de STRUCTURE du type composé.
  const REF_TYPE_MESSAGE = (item) => `Type invalide — attendu : ${item}`;
  const LIST_TYPE_MESSAGE = "Doit être une liste valide.";

  // Primitifs acceptés comme item de `list[item]` — MÊME table que
  // parser.PRIMITIVE_TYPES moins `list` (un `list[list]` n'a pas de sens
  // à valider par item ici, et `list[dict]` reste un item objet possible).
  const PRIMITIVE_ITEMS = { string: true, int: true, float: true, bool: true, dict: true };

  function primitiveOk(value, item) {
    if (item === "string") return typeof value === "string";
    if (item === "bool") return typeof value === "boolean";
    if (item === "int") return typeof value === "number" && Number.isInteger(value);
    if (item === "float") return typeof value === "number" && Number.isFinite(value);
    if (item === "dict") return typeof value === "object" && value !== null && !Array.isArray(value);
    return true; // item inconnu -> passthrough (le serveur reste la vérité)
  }

  function resolveSchema(name) {
    const schema = window.XWEB_SCHEMAS && window.XWEB_SCHEMAS[name];
    if (!schema) {
      console.error(
        `[xweb] validators.js: schéma "${name}" absent de window.XWEB_SCHEMAS ` +
          `(un channel { validate: } ou endpoint { receive.schema: } doit l'exporter sur cette page)`,
      );
    }
    return schema;
  }

  function validateField(value, rules) {
    const errors = [];
    for (const [ruleName, args] of rules) {
      const check = RULES[ruleName];
      if (!check) {
        console.error(`[xweb] validators.js: règle de validation inconnue côté client: "${ruleName}"`);
        errors.push(`Règle de validation inconnue: ${ruleName}`);
        continue;
      }
      if (!check(value, args)) {
        errors.push(MESSAGES[ruleName](args));
      }
    }
    return errors;
  }

  // Valide UN objet contre un schéma, erreurs préfixées par *prefix* ("" au
  // niveau racine, "address." / "phones[0]." pour les champs composés).
  // Gère les pseudo-règles de composition `ref`/`list` en plus de RULES.
  function validateObject(schema, data, prefix, errors) {
    for (const field of Object.keys(schema)) {
      const rules = schema[field];
      const key = prefix + field;
      const value = data == null ? undefined : data[field];
      const fieldErrors = [];

      for (const [ruleName, args] of rules) {
        if (ruleName === "ref") {
          if (value === null || value === undefined) continue; // oblig. géré par required
          if (typeof value !== "object" || Array.isArray(value)) {
            fieldErrors.push(REF_TYPE_MESSAGE(args[0]));
          } else {
            const refSchema = resolveSchema(args[0]);
            if (refSchema) validateObject(refSchema, value, key + ".", errors);
            else fieldErrors.push(`Schéma imbriqué introuvable: ${args[0]}`);
          }
        } else if (ruleName === "list") {
          if (value === null || value === undefined) continue; // oblig. géré par required
          if (!Array.isArray(value)) {
            fieldErrors.push(LIST_TYPE_MESSAGE);
            continue;
          }
          const item = args[0];
          if (item in PRIMITIVE_ITEMS) {
            for (let i = 0; i < value.length; i++) {
              if (!primitiveOk(value[i], item)) {
                errors[key + "[" + i + "]"] = [REF_TYPE_MESSAGE(`list[${item}]`)];
              }
            }
          } else {
            const refSchema = resolveSchema(item);
            if (refSchema) {
              for (let i = 0; i < value.length; i++) {
                validateObject(refSchema, value[i], key + "[" + i + "].", errors);
              }
            } else {
              fieldErrors.push(`Schéma imbriqué introuvable: ${item}`);
            }
          }
        } else {
          fieldErrors.push(...validateField(value, [[ruleName, args]]));
        }
      }
      if (fieldErrors.length > 0) {
        errors[key] = (errors[key] || []).concat(fieldErrors);
      }
    }
  }

  function XwebValidate(schemaName, data, opts) {
    const schema = resolveSchema(schemaName);
    if (!schema) return { valid: false, errors: {} };
    const errors = {};
    if (opts && opts.list === true) {
      // `receive { list: true }` — la réponse JSON attendue est un tableau.
      if (!Array.isArray(data)) {
        return { valid: false, errors: { __root__: [LIST_TYPE_MESSAGE] } };
      }
      for (let i = 0; i < data.length; i++) {
        validateObject(schema, data[i], i + ".", errors);
      }
    } else {
      validateObject(schema, data, "", errors);
    }
    return { valid: Object.keys(errors).length === 0, errors };
  }

  window.XwebValidate = XwebValidate;
})();

File: xdsl/validators.py
"""Moteur de validation déclarative pour les endpoints DSL.

Chaque règle de validation est un objet ``Validator`` qui prend une
valeur et retourne ``True`` si elle est valide, ``False`` sinon. Le
message d'erreur est en français (public cible : développeurs francophones
utilisant xweb).

Usage typique (via le compiler) :
    rules = [Required(), MinLength(3), Email()]
    errors = validate_all("ab", rules)  # → ["Ce champ est obligatoire", "Minimum 3 caractères"]
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any


# ---------------------------------------------------------------------------
# Base
# ---------------------------------------------------------------------------

@dataclass
class ValidationResult:
    """Résultat de la validation d'une valeur."""
    valid: bool
    errors: list[str] = field(default_factory=list)


class Validator:
    """Base abstraite pour toutes les règles de validation."""

    name: str = ""
    message: str = ""

    def check(self, value: Any) -> bool:
        """Retourne True si la valeur passe la règle."""
        raise NotImplementedError

    def __call__(self, value: Any) -> ValidationResult:
        if self.check(value):
            return ValidationResult(valid=True)
        return ValidationResult(valid=False, errors=[self.message])


# ---------------------------------------------------------------------------
# Règles intégrées
# ---------------------------------------------------------------------------

class Required(Validator):
    """Le champ doit être présent et non vide."""
    name = "required"
    message = "Ce champ est obligatoire"

    def check(self, value: Any) -> bool:
        if value is None:
            return False
        if isinstance(value, str) and not value.strip():
            return False
        return True


class MinLength(Validator):
    """Longueur minimale d'une chaîne."""
    name = "min_length"

    def __init__(self, min_len: int) -> None:
        self.min_len = min_len
        self.message = f"Minimum {min_len} caractère{'s' if min_len > 1 else ''}"

    def check(self, value: Any) -> bool:
        if not isinstance(value, str):
            return True  # ne s'applique qu'aux strings
        return len(value) >= self.min_len


class MaxLength(Validator):
    """Longueur maximale d'une chaîne."""
    name = "max_length"

    def __init__(self, max_len: int) -> None:
        self.max_len = max_len
        self.message = f"Maximum {max_len} caractère{'s' if max_len > 1 else ''}"

    def check(self, value: Any) -> bool:
        if not isinstance(value, str):
            return True
        return len(value) <= self.max_len


class MinValue(Validator):
    """Valeur minimale (nombre)."""
    name = "min"

    def __init__(self, min_val: float) -> None:
        self.min_val = min_val
        self.message = f"La valeur minimale est {min_val}"

    def check(self, value: Any) -> bool:
        try:
            return float(value) >= self.min_val
        except (TypeError, ValueError):
            return True


class MaxValue(Validator):
    """Valeur maximale (nombre)."""
    name = "max"

    def __init__(self, max_val: float) -> None:
        self.max_val = max_val
        self.message = f"La valeur maximale est {max_val}"

    def check(self, value: Any) -> bool:
        try:
            return float(value) <= self.max_val
        except (TypeError, ValueError):
            return True


class Email(Validator):
    """Format email valide."""
    name = "email"
    message = "Adresse email invalide"
    _pattern = re.compile(r"^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$")

    def check(self, value: Any) -> bool:
        if not isinstance(value, str):
            return True
        return bool(self._pattern.match(value))


class Pattern(Validator):
    """Expression régulière personnalisée."""
    name = "pattern"

    def __init__(self, regex: str, message: str = "") -> None:
        self.regex = re.compile(regex)
        self.message = message or f"La valeur ne correspond pas au format attendu"

    def check(self, value: Any) -> bool:
        if not isinstance(value, str):
            return True
        return bool(self.regex.match(value))


class In(Validator):
    """La valeur doit être dans une liste de valeurs autorisées."""
    name = "in"

    def __init__(self, allowed: list[str]) -> None:
        self.allowed = allowed
        self.message = f"Valeur invalide — autorisées : {', '.join(allowed)}"

    def check(self, value: Any) -> bool:
        return str(value) in self.allowed


class OneOf(Validator):
    """La valeur doit être l'un des types listés (pour validation type brute)."""
    name = "one_of"

    def __init__(self, types: list[str]) -> None:
        self.types = types
        self.message = f"Type invalide — attendu : {', '.join(types)}"

    def check(self, value: Any) -> bool:
        type_map = {
            "string": str, "int": int, "float": (int, float),
            "bool": bool, "list": list, "dict": dict,
        }
        for t in self.types:
            expected = type_map.get(t)
            if expected and isinstance(value, expected):
                return True
        return False


# ---------------------------------------------------------------------------
# Registre de rules pour le DSL
# ---------------------------------------------------------------------------

# Noms DSL → classe Validator (pour le parser/compiler)
BUILTIN_VALIDATORS: dict[str, type[Validator]] = {
    "required": Required,
    "min_length": MinLength,
    "max_length": MaxLength,
    "min": MinValue,
    "max": MaxValue,
    "email": Email,
    "pattern": Pattern,
    "in": In,
}


def make_validator(name: str, *args: Any) -> Validator:
    """Crée un validator à partir de son nom DSL et de ses arguments.

    Exemples:
        make_validator("required") → Required()
        make_validator("min_length", 3) → MinLength(3)
        make_validator("email") → Email()
    """
    cls = BUILTIN_VALIDATORS.get(name)
    if cls is None:
        raise ValueError(f"Règle de validation inconnue : {name!r}")
    try:
        return cls(*args)
    except TypeError:
        raise ValueError(f"Argument(s) invalide(s) pour la règle {name!r}")


# ---------------------------------------------------------------------------
# API publique
# ---------------------------------------------------------------------------

def validate_all(value: Any, rules: list[Validator]) -> list[str]:
    """Valide une valeur contre une liste de règles.
    Retourne la liste des messages d'erreur (vide si valide)."""
    errors: list[str] = []
    for rule in rules:
        result = rule(value)
        if not result.valid:
            errors.extend(result.errors)
    return errors


def validate_dict(data: dict[str, Any], schema: dict[str, list[Validator]]) -> dict[str, list[str]]:
    """Valide un dict de données contre un schéma de validation.
    Retourne un dict {field: [errors]} — vide si tout est valide."""
    errors: dict[str, list[str]] = {}
    for field_name, rules in schema.items():
        value = data.get(field_name)
        field_errors = validate_all(value, rules)
        if field_errors:
            errors[field_name] = field_errors
    return errors
